File: leetcode/medium/1381. Design a Stack With Increment Operation.ts
class CustomStack {
  protected storage: number[];
  protected max_storage: number;
  protected top: number;
  constructor(maxSize: number) {
    this.top = -1;
    this.max_storage = maxSize;
    this.storage = new Array(this.max_storage);
  }

  push(x: number): void {
    if (this.top < this.max_storage - 1) {
      this.storage[this.top + 1] = x;
      this.top += 1;
    }
  }

  pop(): number {
    if (this.top === -1) return -1;
    const num: number = this.storage[this.top];
    delete this.storage[this.top];
    this.top -= 1;
    return num;
  }

  increment(k: number, val: number): void {
    for (let i = 0; i < k; i += 1) {
      if (i > this.top) break;
      this.storage[i] += val;
    }
  }
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * var obj = new CustomStack(maxSize)
 * obj.push(x)
 * var param_2 = obj.pop()
 * obj.increment(k,val)
 */

File: leetcode/medium/1381. Design a Stack With Increment Operation.rs
struct CustomStack {
    storage: Vec<i32>,
    max_size: i32,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl CustomStack {

    fn new(maxSize: i32) -> Self {
        CustomStack{
            storage: vec![], max_size: maxSize
        }
    }
    
    fn push(&mut self, x: i32) {
        if self.storage.len() < self.max_size as usize{
            self.storage.push(x);
        }
    }
    
    fn pop(&mut self) -> i32 {
        if self.storage.len() == 0{
            return -1;
        }
        let top_num: i32 = self.storage[self.storage.len()-1];
        self.storage.pop();
        return top_num;
    }
    
    fn increment(&mut self, k: i32, val: i32) {
        for i in 0..self.storage.len(){
            if i >= k as usize{
                break;
            }
            self.storage[i] += val;
        }
    }
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * let obj = CustomStack::new(maxSize);
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * obj.increment(k, val);
 */